onnxruntime/core/framework/prepacked_weights_container.h

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/framework/buffer_deleter.h"
+#include "core/framework/tensor_shape.h"
+#include "core/framework/allocator.h"
+#include <unordered_map>
+#include <string>
+
+namespace onnxruntime {
+
+struct PrepackedWeight final {
+  // Some weights may be associated with multiple pre-packed buffers.
+  // Hence we hold them in containers. It is upto the developer implementing each PrePack()
+  // method to define what gets stored in which position of the containers.
+
+  // NOTE: Not all fields may be filled in and not all containers will have the same number of elements
+  // It is upto the developer of the kernel to decide which fields to cache for re-use.
+
+  std::vector<std::unique_ptr<void, BufferDeleter>> buffers_;  // cache pre-packed buffers associated with the kernel
+  std::vector<size_t> weights_sizes_;                          // cache sizes associated with pre-packed buffers
+  std::vector<TensorShape> shapes_;                            // cache tensor shapes associates with pre-packed buffers
+  std::vector<bool> flags_;                                    // cache some flags associated with the pre-packed buffers
+
+  bool is_filled_ = false;  // By default, an instance of this class is "unfilled"
+};
+
+// TODO: Make this class thread-safe ?
+class PrepackedWeightsContainer final {
+ public:
+  PrepackedWeightsContainer() {
+  }
+
+  ~PrepackedWeightsContainer() = default;
+
+  AllocatorPtr GetAllocator(const std::string& device_name);
+
+  const PrepackedWeight& GetCachedWeight(const std::string& initializer_name);
```

Comment:
One corner case is that one initializer may be used by 2 nodes with different op types. In this case, the packing format can be different.

Reply:
Should we allow this? I think I remember asking you this before and you'd said this is not a use case.

Reply:
Should the key in the cache just be op_type + initializer name then ?

Reply:
The more I thought about it - I think it may be possible to have an initializer used by multiple nodes of different op types. So to avoid any potential issues - I added the op type to the container lookup key. So, we will only use the cached pre-packed weight if the shared constant initializer is being used by multiple nodes of the same op type.